src/main/scala/workflow/NodeOptimizationRule.scala

```diff
@@ -5,96 +5,341 @@ import org.apache.spark.rdd.RDD
+  case class OptimizationState(
+    optimizedInstructions: Seq[Instruction],
+    unoptimizedInstructionIndexToOptimizedIndex: Int => Int,
+    registers: Map[Int, InstructionOutput],
+    numPerPartitionPerNode: Map[Int, Map[Int, Int]]
+  )
+
+  /**
+   * Get the set of instructions that we need to check for optimizations at.
+   *
+   * This is any instruction that either applies an optimizable transformer,
+   * or fits an optimizable estimator,
+   * and is not supposed to be executed on test data.
+   */
+  private def getInstructionsToOptimize(instructions: Seq[Instruction]): Set[Int] = {
+    instructions.zipWithIndex.collect {
+      case (TransformerApplyNode(tIndex, _), index)
+        if instructions(tIndex).isInstanceOf[OptimizableTransformer[_, _]] => index
+
+      case (EstimatorFitNode(estIndex, _), index)
+        if instructions(estIndex).isInstanceOf[OptimizableEstimator[_, _]] => index
+
+      case (EstimatorFitNode(estIndex, _), index)
+        if instructions(estIndex).isInstanceOf[OptimizableLabelEstimator[_, _, _]] => index
+
+    }.toSet -- WorkflowUtils.getChildren(Pipeline.SOURCE, instructions)
+  }
+
+  /**
+   * Execute the instruction at the given index, and output the new state of the optimization process
+   *
+   * @param optimizationState
+   * @param instruction
+   * @param index
+   * @return
+   */
+  private def executeInstruction(optimizationState: OptimizationState, instruction: Instruction, index: Int)
+  : OptimizationState = {
+    // Get the dependencies from the registers and run the instruction on them (caching RDD outputs)
+    val registers = optimizationState.registers
+    val deps = instruction.getDependencies.map(registers)
+    val instructionOutput = instruction.execute(deps) match {
+      case RDDOutput(rdd) => RDDOutput(rdd.cache())
+      case out => out
+    }
+    val newRegisters = registers + (index -> instructionOutput)
+
+    // If any of the dependencies is an RDD and has a numPerPartition, make sure to feed it forward
+    val numPerPartitionPerNode = optimizationState.numPerPartitionPerNode
+    val numPerPartition = instruction.getDependencies.collectFirst {
+      case dep if numPerPartitionPerNode.contains(dep) => numPerPartitionPerNode(dep)
+    }
+    val newNumPerPartitionPerNode = if (numPerPartition.nonEmpty) {
+      numPerPartitionPerNode + (index -> numPerPartition.get)
+    } else {
+      numPerPartitionPerNode
+    }
+
+    // Create the new state
+    optimizationState.copy(registers = newRegisters, numPerPartitionPerNode = newNumPerPartitionPerNode)
+  }
+
+  /**
+   * Optimizes the transformer being applied by the instruction at the given index.
+   * Outputs a new optimization state.
+   *
+   * @param os The previous optimization state
+   * @param transformerApplyNode
+   * @param index
+   * @return
+   */
+  private def optimizeTransformer(os: OptimizationState, transformerApplyNode: TransformerApplyNode, index: Int)
```

Comment:
Yikes - os is a bad name for this. how about `optState` or something? `os.copy` made me think that we were copying files and I thought something was very wrong..